pages/api/assets.py

```diff
@@ -138,91 +137,103 @@ def post(self, request, *args, **kwargs):
+            prefix = prefix[len(media_prefix) + 1:]
+        storage_key_name = "%s%s%s" % (prefix,
+                                       hashlib.sha256(uploaded_file.read()).hexdigest(), ext)
+
+        dst_key_name = "%s/%s" % (media_prefix, storage_key_name)
+        LOGGER.info("S3 bucket %s: copy %s to %s",
+                    bucket_name, src_key_name, dst_key_name)
+        storage = get_default_storage(request)
+        if is_public_asset:
+            extra_args = {'ACL': "public-read"}
+        else:
+            extra_args = {
+                'ServerSideEncryption': settings.AWS_SERVER_SIDE_ENCRYPTION}
+        if ext in ['.pdf']:
+            extra_args.update({'ContentType': 'application/pdf'})
+        elif ext in ['.jpg']:
+            extra_args.update({'ContentType': 'image/jpeg'})
+        elif ext in ['.png']:
+            extra_args.update({'ContentType': 'image/png'})
+        s3_client.copy({'Bucket': bucket_name, 'Key': src_key_name},
+                       bucket_name, dst_key_name, ExtraArgs=extra_args)
+        # XXX still can't figure out why we get a permission denied on DeleteObject.
+        #            s3_client.delete_object(Bucket=bucket_name, Key=src_key_name)
+        location = storage.url(storage_key_name)
+
+    elif 'file' in request.data:
+        uploaded_file = request.data['file']
+        if content_type:
+            # We optionally force the content_type because S3Store uses
+            # mimetypes.guess and surprisingly it doesn't get it correct
+            # for 'text/css'.
+            uploaded_file.content_type = content_type
+        sha1 = hashlib.sha1(uploaded_file.read()).hexdigest()
+
+        # Store filenames with forward slashes, even on Windows
+        filename = force_str(uploaded_file.name.replace('\\', '/'))
+        sha1_filename = sha1 + os.path.splitext(filename)[1]
+        storage = get_default_storage(request)
+        stored_filename = sha1_filename if store_hash else filename
+        if not is_public_asset:
+            stored_filename = '/'.join(
+                [str(account), stored_filename])
+
+        LOGGER.debug("upload %s to %s", filename, stored_filename)
+        if storage.exists(stored_filename) and replace_stored:
+            storage.delete(stored_filename)
+        storage.save(stored_filename, uploaded_file)
+        response_status = status.HTTP_201_CREATED
+        location = storage.url(stored_filename)
+
+    else:
+        raise ValidationError({'detail':
+           _("Either 'location' or 'file' must be specified.")})
+
+    if not is_public_asset:
+        path = urlparse(location).path.lstrip(URL_PATH_SEP)
+        if path.startswith(media_prefix):
+            path = path[len(media_prefix):].lstrip(URL_PATH_SEP)
+        location = request.build_absolute_uri(
+            reverse('pages_api_asset', args=(account, path)))
+
+    return ({
+        'location': location,
+        'updated_at': datetime_or_now()},
```

Comment:
I am not sure `'updated_at': datetime_or_now()` is useful. Somehow I feel the `updated_at` should match the file modified date/time.